jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java

```diff
@@ -57,6 +69,8 @@
 
    private AbstractJCacheListenerAdapter<K,V> listenerAdapter;
 
+   private ConcurrentMap<List<Object>, Queue<CountDownLatch>> latchesByEventSource = new ConcurrentHashMap<>();
```

Comment:
I agree regarding the List as a key.  The main issue here is that if you modified that list after you inserted you have changed the hashCode and equality which means that no one would be able to find this entry again.